stl/inc/ratio

```diff
@@ -19,59 +20,54 @@ _STL_DISABLE_CLANG_WARNINGS
 #undef new
 
 _STD_BEGIN
-template <intmax_t _Val>
-struct _Abs : integral_constant<intmax_t, (_Val < 0 ? -_Val : _Val)> {}; // computes absolute value of _Val
-
-template <intmax_t _Ax, intmax_t _Bx, bool _Sfinae = false,
-    bool _Good = (_Abs<_Ax>::value <= INTMAX_MAX / (_Bx == 0 ? 1 : _Abs<_Bx>::value))>
-struct _Safe_mult : integral_constant<intmax_t, _Ax * _Bx> {}; // computes _Ax * _Bx without overflow
+_NODISCARD constexpr intmax_t _Abs(intmax_t _Val) { // computes absolute value sof _Val
+    return (_Val < 0 ? -_Val : _Val);
+}
 
-template <intmax_t _Ax, intmax_t _Bx, bool _Sfinae>
-struct _Safe_mult<_Ax, _Bx, _Sfinae, false> { // _Ax * _Bx would overflow
-    static_assert(_Sfinae, "integer arithmetic overflow");
-};
+_NODISCARD constexpr intmax_t _Safe_mult(intmax_t _Ax, intmax_t _Bx) { // computes _Ax * _Bx without overflow
+    if (_Abs(_Ax) > INTMAX_MAX / (_Bx == 0 ? 1 : _Abs(_Bx))) {
+        _CSTD abort(); // assert does not work in NDEBUG
+    }
 
-template <intmax_t _Val>
-struct _Sign_of : integral_constant<intmax_t, (_Val < 0 ? -1 : 1)> {}; // computes sign of _Val
+    return _Ax * _Bx;
+}
 
-template <intmax_t _Ax, intmax_t _Bx, bool _Good, bool _Also_good>
-struct _Safe_addX : integral_constant<intmax_t, _Ax + _Bx> {}; // computes _Ax + _Bx without overflow
+_NODISCARD constexpr intmax_t _Sign_of(intmax_t _Val) { // computes sign of _Val
+    return (_Val < 0 ? -1 : 1);
+}
 
-template <intmax_t _Ax, intmax_t _Bx>
-struct _Safe_addX<_Ax, _Bx, false, false> { // _Ax + _Bx would overflow
-    static_assert(_Always_false<_Safe_addX>, "integer arithmetic overflow");
-};
+_NODISCARD constexpr intmax_t _Safe_add(intmax_t _Ax, intmax_t _Bx) { // computes _Ax + _Bx without overflow
+    if (_Sign_of(_Ax) == _Sign_of(_Bx) && (_Abs(_Ax) > INTMAX_MAX - _Abs(_Bx))) {
+        _CSTD abort(); // assert does not work in NDEBUG
+    }
 
-template <intmax_t _Ax, intmax_t _Bx>
-struct _Safe_add : _Safe_addX<_Ax, _Bx, _Sign_of<_Ax>::value != _Sign_of<_Bx>::value,
-                       (_Abs<_Ax>::value <= INTMAX_MAX - _Abs<_Bx>::value)>::type {
-    // computes _Ax + _Bx, forbids overflow
-};
+    return _Ax + _Bx;
+}
 
-template <intmax_t _Ax, intmax_t _Bx>
-struct _GcdX : _GcdX<_Bx, _Ax % _Bx>::type {}; // computes GCD of _Ax and _Bx
+_NODISCARD constexpr intmax_t _GcdX(intmax_t _Ax, intmax_t _Bx) { // computes GCD of abs(_Ax) and abs(_Bx)
+    if (_Bx == 0) {
+        return _Ax; // computes GCD of _Ax and 0
+    }
 
-template <intmax_t _Ax>
-struct _GcdX<_Ax, 0> : integral_constant<intmax_t, _Ax> {}; // computes GCD of _Ax and 0
+    return _GcdX(_Bx, _Ax % _Bx);
+}
 
-template <intmax_t _Ax, intmax_t _Bx>
-struct _Gcd : _GcdX<_Abs<_Ax>::value, _Abs<_Bx>::value>::type {}; // computes GCD of abs(_Ax) and abs(_Bx)
+_NODISCARD constexpr intmax_t _Gcd(intmax_t _Ax, intmax_t _Bx) { // computes GCD of abs(_Ax) and abs(_Bx)

```

Comment:
Please fuse this with `_GcdX`, the whole point in reducing the number of functions

Reply:
Done